### ISSUE TITLE
clang-cl, mismatch MSVC cl about `error: redefinition of 'XXX' as different kind of symbol`

### ISSUE THREAD

**Issue by lygstate** [llvm/llvm-project#95681]
Code to reproduce the issue

```cpp
#include <windows.h>
DPAPI_IMP
BOOL
WINAPI
wpCryptProtectMemory(
    _Inout_         LPVOID          pDataIn,             // in out data to encrypt
    _In_            DWORD           cbDataIn,            // multiple of CRYPTPROTECTMEMORY_BLOCK_SIZE
    _In_            DWORD           dwFlags
    )
{
    return TRUE;
}

namespace test
{
extern "C" __declspec(selectany) void const *const CryptProtectMemory = nullptr;
}

int main()
{
    return 0;
}

```

With cl.exe, it's compiles fine.
With clang-cl.exe, it's comes with
```
C:\work\study\runtimes\win-polyfill>clang-cl test.cpp
test.cpp(16,52): error: redefinition of 'CryptProtectMemory' as different kind of symbol
   16 | extern "C" __declspec(selectany) void const *const CryptProtectMemory = nullptr;
      |                                                    ^
C:\Program Files (x86)\Windows Kits\10\\include\10.0.26100.0\\um\dpapi.h(312,1): note: previous definition is here
  312 | CryptProtectMemory(
      | ^
1 error generated.
```

**Comment by zufuliu** [comment]
Similar to issue #58783.

**Comment by lygstate** [comment]
> Similar to issue #58783.

yeap, somewhat different, this have namespace and for C++ only, without the namespace, MSVC also have error message

**Comment by AaronBallman** [comment]
I think Clang is correct to reject that code, per https://eel.is/c++draft/dcl.link#7 and https://eel.is/c++draft/basic.scope#scope-6: https://godbolt.org/z/nPc5451fv -- they don't declare the same entity and the names conflict.

CC @rnk @zmodem for opinions on whether this is behavior we should try to implement for compatibility.

**Comment by lygstate** [comment]
> I think Clang is correct to reject that code, per https://eel.is/c++draft/dcl.link#7 and https://eel.is/c++draft/basic.scope#scope-6: https://godbolt.org/z/nPc5451fv -- they don't declare the same entity and the names conflict.
> 
> CC @rnk @zmodem for opinions on whether this is behavior we should try to implement for compatibility. My inclination is to not support this construct unless there's a strong need; this feels like a bug more than an intentional behavior (unless I'm missing something).

I need this to export symbols to generate symbols to override windows sdk API, This is something I needed. I think this can be treat as warning instead of error(so that the warning can be disabled)

**Comment by zmodem** [comment]
> I think Clang is correct to reject that code, per https://eel.is/c++draft/dcl.link#7 and https://eel.is/c++draft/basic.scope#scope-6: https://godbolt.org/z/nPc5451fv -- they don't declare the same entity and the names conflict.
> 
> CC @rnk @zmodem for opinions on whether this is behavior we should try to implement for compatibility. My inclination is to not support this construct unless there's a strong need; this feels like a bug more than an intentional behavior (unless I'm missing something).

I agree with your inclination, we shouldn't do compatibility workarounds unless it's for code which a user cannot control such as Windows system headers. That doesn't seem to be the case here.


> I need this to export symbols to generate symbols to override windows sdk API, This is something I needed. I think this can be treat as warning instead of error(so that the warning can be disabled)

If you just need to generate a symbol with the same name, can't you do it in a file which doesn't include windows.h?

**Comment by lygstate** [comment]
> > I think Clang is correct to reject that code, per https://eel.is/c++draft/dcl.link#7 and https://eel.is/c++draft/basic.scope#scope-6: https://godbolt.org/z/nPc5451fv -- they don't declare the same entity and the names conflict.
> > CC @rnk @zmodem for opinions on whether this is behavior we should try to implement for compatibility. My inclination is to not support this construct unless there's a strong need; this feels like a bug more than an intentional behavior (unless I'm missing something).
> 
> I agree with your inclination, we shouldn't do compatibility workarounds unless it's for code which a user cannot control such as Windows system headers. That doesn't seem to be the case here.
> 
> > I need this to export symbols to generate symbols to override windows sdk API, This is something I needed. I think this can be treat as warning instead of error(so that the warning can be disabled)
> 
> If you just need to generate a symbol with the same name, can't you do it in a file which doesn't include windows.h?

Yeap, because I am trying to override Windows.h's api, I depends on that :(

**Comment by zufuliu** [comment]
It seems this is same as issue #24569, #29512.
